File: 6.zig-zag-conversion.py
#
# @lc app=leetcode id=6 lang=python3
#
# [6] ZigZag Conversion
#

# Time:  O(n)
# Space: O(1)

# @lc code=start
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        step, zigzag = numRows * 2 - 2, ""
        for i in range(numRows):
            for j in range(i, len(s), step):
                zigzag += s[j]
                if 0 < i < numRows - 1 and j + step - 2 * i < len(s):
                    zigzag += s[j + step - 2 * i]
        return zigzag


# @lc code=end

File: 6.zig-zag-conversion.js
/*
 * @lc app=leetcode id=6 lang=javascript
 *
 * [6] ZigZag Conversion
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (numRows === 1) {
    return s;
  }
  let res = "";
  const step = numRows * 2 - 2;
  for (let i = 0; i < numRows; ++i) {
    for (let j = i; j < s.length; j += step) {
      res += s[j];
      if (0 < i && i < numRows - 1 && j + step - i * 2 < s.length) {
        res += s[j + step - i * 2];
      }
    }
  }
  return res;
};
// @lc code=end
